R/derive_param_clinbenefit.R

```diff
@@ -0,0 +1,256 @@
+                                     source_datasets,
+                                     by_vars = NULL,
+                                     reference_date = TRTSDT,
+                                     ref_start_window = 28,
+                                     set_values_to,
+                                     subject_keys = vars(STUDYID, USUBJID)) {
+
+  # Assertions and quotes
+  assert_vars(by_vars, optional = TRUE)
+  reference_date <- assert_symbol(enquo(reference_date))
+  assert_data_frame(
+    dataset_adsl,
+    required_vars = vars(!!!subject_keys, !!reference_date)
+  )
+  assert_data_frame(
+    dataset,
+    required_vars = vars(!!!by_vars, PARAMCD, AVALC, ADT)
+  )
+
+  filter_source <- assert_filter_cond(
+    enquo(filter_source),
+    optional = FALSE
+  )
+
+  assert_vars(subject_keys)
+  assert_s3_class(source_resp, "date_source")
+  assert_s3_class(source_pd, "date_source")
+  assert_list_of(source_datasets, "data.frame")
+
+  source_names <- names(source_datasets)
+  if (!all(c(source_pd$dataset_name, source_resp$dataset_name) %in% source_names)) {
+    abort(
+      paste0(
+        "The dataset names specified for `source_pd` and `source_resp` must be \n",
+        "included in the list specified for the `source_datasets` parameter.\n",
+        "Following names were provided by `source_datasets`:\n",
+        enumerate(source_names, quote_fun = squote)
+      )
+    )
+  }
+
+  ref_start_window <- assert_integer_scalar(ref_start_window)
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+  assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+
+  # ADSL variables
+
+  adsl_vars <- vars(
+    !!!subject_keys,
+    !!reference_date
+  )
+
+  adsl <- dataset_adsl %>%
+    select(!!!adsl_vars)
+
+  # Get PD date and objective response date
+
+  pd_data <- source_datasets[[source_pd$dataset_name]] %>%
+    filter_if(source_pd$filter) %>%
+    select(!!!subject_keys, !!!by_vars, !!source_pd$date) %>%
+    rename(temp_pd = !!source_pd$date)
+
+  rsp_data <- source_datasets[[source_resp$dataset_name]] %>%
+    filter_if(source_resp$filter) %>%
+    select(!!!subject_keys, !!!by_vars, !!source_resp$date) %>%
+    rename(temp_rs = !!source_resp$date)
+
+  # Look for valid non-PD measurements after window from reference date
+  ovr_data <- filter_pd(
+    dataset = dataset,
+    filter = !!filter_source,
+    source_pd = source_pd,
+    source_datasets = source_datasets
+  )
+
+  ovr_data <- ovr_data %>%
+    left_join(
+      adsl,
+      by = vars2chr(subject_keys)
+    ) %>%
+    left_join(
+      pd_data,
+      by = vars2chr(subject_keys)
+    ) %>%
+    filter(
+      !(AVALC %in% c("NA", "NE", "ND")) & !is.na(AVALC) &
+        ADT >= !!reference_date + days(ref_start_window)
+    ) %>%
+    # Use this approach only for patients that are not already responders
+    anti_join(
+      .,
+      rsp_data,
+      by = vars2chr(subject_keys)
+    ) %>%
```

Comment:
what would our STREAM macro give here @bundfussr out of interest? i'd guess it would use the PR date. i agree with your point though and i would have thought it makes most sense to use the ADT of whichever case comes first. i see this derivation less of a IF _responder_ THEN _use this_ ELSE _check sd more than zzz days_ etc, and more of a check IF EITHER case then it's fine and you're considered to have clinical benefit,

as per my above comment though i don't really find ADT interesting here for most analyses.

Reply:
STREAM would use the PR date (but this may a misinterpretation of the specs). I am not aware of any standard analysis which uses the date of clinical benefit. This may be the reason why nobody complained.

For admiral I would use the date which comes first. Otherwise other companies may not be able to use the function because it deviates from their specs.

Reply:
makes sense to me! @amsmith214 i would also then re-word your header info for this function as currently it reads "Clinical benefit/disease control is **first identified for looking for subjects having objective response**, and then derived...". In reality you check either condition and then just take the first occurring.

Reply:
i agree with the example you pose. i guess in that case i'm confused as to why we need `source_resp` then, since taking the earliest measurement that's not `NE, NA, ND, PD` i think will always be at or before the date specified by `source_resp`.

in general i personally find the structure of the functions so far very confusing

Reply:
For the measurements not NE or PD we take the time window into account. For CR or PR this is not required. The function can also be used for confirmed clinical benefit. In this case the confirmed response parameter would be specified for `source_resp`. Thus we need `source_resp`.

I agree that the functions looks confusing. I am curious about the tester's feedback.

Reply:
@amsmith214 does the first draft ADRS vignette help to understand the flow more and why the functions have been designed as they are? see https://github.com/pharmaverse/admiralonco/blob/10_adrs_vignette_template/vignettes/adrs.Rmd

Reply:
it does help, definitely. i just personally think developing a complete understanding of these functions is going to be onerous for users. for example, i think to really know `derive_param_clinbenefit`, a user would have to understand the various `date_source` objects, `filter_if`, `filter_pd`, and `filter_extreme`, all bespoke to admiral.

i'm not suggesting people learn machine code, or that (e.g.) it's common to know exactly what SAS `proc means` or `proc freq` or even `dplyr::filter` are up to under the hood. i also think the error handling baked into admiral will cut off a lot of potential debug issues at the pass, but when something new goes wrong, i think it could potentially be tough for a user to sort out with the current nesting structure of functions.

Reply:
yeh fair points and this is an age old debate we have on admiral. Stefan's argument is always that our functional documentation should be so clear that users never need to look under the hood at the code. We do have to be careful to try and stay true to https://pharmaverse.github.io/admiral/index.html#simplicity though. This does get challenging when we're dealing with complex endpoints.

Reply:
When debugging admiral functions the nesting did not cause issues for me. But I struggled with the use of `%>%` in our code. You can not see the intermediate datasets in a `%>%` chain and going into one of the functions called in the chain is really hard.
@thomas-neitmann , how do you debug `%>%` chains?

Reply:
If you want to see intermediate datasets than having `%>%` certainly makes things more difficult. Certainly one reason why you should not pipe too many operations one after another. Without going into the code directly and either breaking apart the pipe chain or inserting some `print()` calls in between the only option is to directly `debug()` or `debugonce()` the function in the chain which fails to see the input dataset that is passed onto it.

Reply:
for the {admiralroche} ADaM templates so far i've been breaking the sections down so only a few pipes each time of similar related variables. and each intermediary dataset is uniquely named, as opposed to the {admiral} convention of overwriting same dataset each time, so that its easy to trace through your working. will be interested to see if the users prefer this during testing. i'm not necessarily suggesting the same for function development or the {admiral} templates, but just sharing as an FYI of the Roche ext package thinking. tagging @millerg23 as fyi as he's working on some of these too.